markdown/extensions/codehilite.py

```diff
@@ -200,13 +200,20 @@ def _parseHeader(self):
 class HiliteTreeprocessor(Treeprocessor):
     """ Hilight source code in code blocks. """
 
+    def code_unescape(self, text):
+        """Unescape code."""
+        text = text.replace("&amp;", "&")
+        text = text.replace("&lt;", "<")
+        text = text.replace("&gt;", ">")
+        return text
+
     def run(self, root):
         """ Find code blocks and store in htmlStash. """
         blocks = root.iter('pre')
         for block in blocks:
             if len(block) == 1 and block[0].tag == 'code':
                 code = CodeHilite(
-                    block[0].text,
+                    self.code_unescape(block[0].text),
```

Comment:
Because this is code, it will unescape all `&amp;` `&lt;` and `&gt;` because in the `code` step, we escape all `&`, `<`, and `>`.  It is done in the code step because we want literal representations of anything in the code region. We do not preserve entities, but the characters. The serializer was updated to avoid already escaped entities, code does *not* do this which is why we do it in the code step.

We only do this to code captured via the `code` step which is why we have this occurring in the `HiliteTreeprocessor` and not in the `CodeHilite` class itself.  Fenced code will capture the code and process it immediately, so no unescaping is needing.  Only indented blocs are handled in two phases requiring an escape.

With all of that said, I will gladly add the requested test below.  We can expect that each `&` will be expanded to `&amp;` in the final output regardless of whether it is part of an entity or not as what we are looking for is preserving the characters.

```
    &
    &amp;
    &amp;amp;
```